[c10d] allow sub group to be eagerly inited even if default one is not

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138665
* #138781

Summary:
Currently, eager mode is applied either to all PGs or NONE of them.
There are cases where we don't want to initialize the comms for default
PG, but we still want to initialize the comms for sub PG. Now with a
device_id passed to new group, we can achieve this case
Test Plan:
newly added UT

Tags:

Resolves https://github.com/pytorch/pytorch/issues/137018

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

shuqiangzhang committed:
[c10d] allow sub group to be eagerly inited even if default one is not
Summary:
Currently, eager mode is applied either to all PGs or NONE of them.
There are cases where we don't want to initialize the comms for default
PG, but we still want to initialize the comms for sub PG. Now with a
device_id passed to new group, we can achieve this case
Test Plan:
newly added UT

Tags:

[ghstack-poisoned]

kwen2501 commented:
I agree with the addition of `device_id` to the `new_group` API.
(I wonder if we should start calling the kwarg `device` or `device_id` -- it is actually a `torch.device`, rather than an id / index. Although, `device_id` makes the API aligned with `init_process_group`, so I am undecided.)

kwen2501 commented:
Does this PR need to depend on https://github.com/pytorch/pytorch/pull/138518?

shuqiangzhang commented:
> Does this PR need to depend on #138518?

Yes. otherwise, eager mode is coupled with split logic, aka, if subgroup is eager inited, it has to use split, which requires default PG to also be eager inited, which is against the intention of this PR

shuqiangzhang commented:
> I agree with the addition of `device_id` to the `new_group` API. (I wonder if we should start calling the kwarg `device` instead of `device_id` -- it is actually a `torch.device`, rather than an id / index. Although, `device_id` makes the API aligned with `init_process_group`, so I am undecided.)

For the purpose of this PR, let's keep it consistent with the naming of init_process_group API? We could re-name all of them in other PRs.

kwen2501 commented:
> if subgroup is eager inited, it has to use split

I wonder if that's true, a subgroup can be eagerly inited the same way as a default group is eagerly inited, without using split.

kwen2501 commented:
> We could re-name all of them in other PRs.

Once added this is a public argument. So there will be deprecation consequence if we'd like to change it in future.

kwen2501 commented:
> let's keep it consistent with the naming of init_process_group API

Okay let's do that.

shuqiangzhang committed:
Update on "[c10d] allow sub group to be eagerly inited even if default one is not"


Summary:
Currently, eager mode is applied either to all PGs or NONE of them.
There are cases where we don't want to initialize the comms for default
PG, but we still want to initialize the comms for sub PG. Now with a
device_id passed to new group, we can achieve this case
Test Plan:
newly added UT

Tags:

Resolves https://github.com/pytorch/pytorch/issues/137018

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4774,6 +4775,11 @@ def new_group(
             in that non-member ranks don't need to call into API and don't
             join the barrier.
         group_desc (str, optional): a string to describe the process group.
+        device_id (torch.device, optional): a single, specific device
+            to "bind" this process to, allowing for backend-specific
+            optimizations.  only under NCCL: the communicator is immediately formed
+            (calling``ncclCommInit*`` immediately rather than the normal lazy
+            call)
```
This piece of documentation seems to come from `init_process_group`. I wonder if we could update them now that our view about `device_id` is clearer? For example:
```
a single, specific device to "bind" this process to. The `new_group` call will try to initialize a communication backend for the device if this field is given.
```

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -4818,7 +4826,8 @@ def _new_group_with_tag(
 
     default_pg = _get_default_group()
-    device_id = default_pg.bound_device_id
+    if device_id is None:
+        device_id = default_pg.bound_device_id
```
the current logic seems fine. Shall we also do a check here to make sure the given `device_id` is the same as `default_pg.bound_device_id` if they are both not `None`?

kwen2501 commented on torch/csrc/distributed/c10d/Backend.hpp:
```diff
@@ -386,6 +386,12 @@ class TORCH_API Backend : public torch::CustomClassHolder {
     // backends may perform
   }
 
+  // whether the backend is fully initialized, e.g., for NCCL, if the NCCL comms
+  // are fully initialized and ready to use.
+  virtual bool isInitialized() {
+    return false;
+  }
+
```
Could it be a bit unsafe to assume a default value here?
Maybe we should throw an unimplemented error here and force backends to implement it? But it will induce more work, including contacting 3rd-party backends to implement this.

shuqiangzhang replied:
If this is problemm We could move this interface to nccl backend only, similar to other APIs such as bound_device_id or abort

shuqiangzhang committed:
Update on "[c10d] allow sub group to be eagerly inited even if default one is not"


Summary:
Currently, eager mode is applied either to all PGs or NONE of them.
There are cases where we don't want to initialize the comms for default
PG, but we still want to initialize the comms for sub PG. Now with a
device_id passed to new group, we can achieve this case
Test Plan:
newly added UT

Tags:

Resolves https://github.com/pytorch/pytorch/issues/137018

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -852,6 +852,30 @@ def test_comm_split_subgroup(self):
+
+        tensor = torch.full((1,), self.rank).cuda(device)
+        new_group = c10d.new_group([0, 1], device_id=device)
+        self.assertEqual(backend.comm_split_count(), 0)
```
nit: I propose we stop using `comm_split_count` for testing.

shuqiangzhang replied:
This one is widely used on split related tests, any alternatives?

kwen2501 replied:
No alternatives. Just proposing that we stop using it.

shuqiangzhang replied:
Then, not using it is basically equivalent to removing valid python tests and safety guards to future code changes. Unless we have an alternative, e.g., fully trusted c++ tests (but still missing the e2e py test)

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -852,6 +852,30 @@ def test_comm_split_subgroup(self):
+        new_backend = new_group._get_backend(torch.device(device))
+        self.assertEqual(new_backend._is_initialized(), True)
+        dist.broadcast(tensor, 0, group=new_group)
+        self.assertEqual(new_backend.comm_split_count(), 0)
```
Same here

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -852,6 +852,30 @@ def test_comm_split_subgroup(self):
+        dist.broadcast(tensor, 0, group=new_group)
+        self.assertEqual(new_backend.comm_split_count(), 0)
+        self.assertEqual(backend._is_initialized(), False)
+        torch.cuda.synchronize()
```
nit: is this synchronize necessary?

shuqiangzhang replied:
Yes, because things could be aborted before collective is completed

kwen2501 reviewed (approved):
LGTM overall. The comments are minor.

shuqiangzhang commented:
@pytorchbot merge -f "no failures"
